internal/tests/serialization/set.go

```diff
@@ -0,0 +1,212 @@
+package serialization
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime/debug"
+	"testing"
+)
+
+// Set is a tool for generating test cases of marshal and unmarshall funcs.
+// For cases when the function should no error,
+// marshaled data from Set.Values should be equal with Set.Data,
+// unmarshalled value from Set.Data should be equal with Set.Values.
+type Set struct {
+	Data   []byte
+	Values []interface{}
+
+	BrokenMarshalTypes   []reflect.Type
+	BrokenUnmarshalTypes []reflect.Type
+}
+
+func (s Set) RunTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
+	if name == "" {
+		t.Fatal("name should be provided")
+	}
+
+	t.Run(name, func(t *testing.T) {
+		for i := range s.Values {
+			val := s.Values[i]
+
+			t.Run(fmt.Sprintf("%T", val), func(t *testing.T) {
+				if marshal != nil {
+					s.runMarshalTest(t, marshal, val)
+				}
+
+				if unmarshal != nil {
+					if rt := reflect.TypeOf(val); rt.Kind() != reflect.Ptr {
+						unmarshalIn := newRef(val)
+						s.runUnmarshalTest("unmarshal", t, unmarshal, val, unmarshalIn)
+					} else {
+						// Test unmarshal to (*type)(nil)
+						unmarshalIn := newRef(val)
+						s.runUnmarshalTest("unmarshal**nil", t, unmarshal, val, unmarshalIn)
+
+						// Test unmarshal to &type{}
+						unmarshalInZero := newRefToZero(val)
+						s.runUnmarshalTest("unmarshal**zero", t, unmarshal, val, unmarshalInZero)
+					}
+				}
+			})
+		}
+	})
+}
+
+func (s Set) RunCorruptTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
```

Comment:
Since you introduced a new method, now you need to explain what is the purpose of `RunCorruptTest`, outlining difference between `RunTest`

Reply:
Done